src/pipx/commands/common.py

```diff
@@ -148,7 +149,8 @@ def get_package_summary(
             f"       Please uninstall and install this package, or reinstall-all to fix."
         )
 
-    package_metadata = venv.package_metadata[package]
+    package_metadata = venv.get_package_metadata(package)
```

Comment:
I think instead of these two line changes, we need to go a little farther back and change starting line 141:
```
if package is None:
    package = path.name
```
to instead be
```
if package is None:
  package = venv.package_metadata.main_package["package"]
```
Since we have now broken old pipx's one-to-one relationship between path <-> main_package name.

Reply:
And then I think you don't need `venv.get_package_metadata()` and you don't need to search for the package name with its for-loop.

Reply:
I am not sure if this will cover all the cases unfortunately. I had something similar in my first attempt. The issue was upgrade of injected packages that failed in this scenario. Additioanlly, the `get_package_metadata` is also useful if the suffix for installed app packages are enabled int he future. While it is probably not the most efficient method, it does the trick reliably. I am happy to make the suggested changes, and handle the key error explicitly in the upgrade code. Unless there is a cleaner way of doing this that I am missing.

Reply:
Can you explain what's going wrong with upgrade?  I wasn't aware we could upgrade injected packages.

Also, could you explain what you mean by "useful if the suffix for installed app packages are enabled int he future"?

In general I'd like to fix the implicit assumption in the code that pathname==main_package, because ever since we have suffixes that is incorrect.

Reply:
I am purely going by the possible code-paths here. It might be that I am misunderstanding what `pipx` wants to support. In theory any package in a venv ca be updated based on this signature.

https://github.com/pipxproject/pipx/blob/20238b1153eb5ade0eb851511dc2cf51c7eac4eb/src/pipx/commands/upgrade.py#L15-L23

Replacing the following with `package_metadata = venv.pipx_metadata.main_package` seems incorrect.

https://github.com/pipxproject/pipx/blob/20238b1153eb5ade0eb851511dc2cf51c7eac4eb/src/pipx/commands/upgrade.py#L42

And following that there is also this call, which I expect is intended to expose apps (my understanding of the pipx app concept might be incomplete) from a pipx environment. 

https://github.com/pipxproject/pipx/blob/20238b1153eb5ade0eb851511dc2cf51c7eac4eb/src/pipx/commands/upgrade.py#L70-L72

Since each of the packages now have a `suffix` property, in theory each exposed app can use a suffix. So this implies that if required, one can expose each app with its own suffix or an inherited suffix. This might not be what the project wants, and I can appreciate that.

I feel that in order to break that assumption safely old installations in the wild will need to be updated I feel. Otherwise a `pipx` upgrade can break existing environments. Again, a fair few assumptions here.

Reply:
Currently we don't support upgrading anything but the main package (see the line `# TODO 20191026: upgrade injected packages also (Issue #79)`)

> Replacing the following with `package_metadata = venv.pipx_metadata.main_package` seems incorrect.

That would be incorrect.  What I was proposing was to obtain metadata on the main package was:
```
package = venv.pipx_metadata.main_package
package_metadata = venv.package_metadata[package]
```

> Since each of the packages now have a suffix property, in theory each exposed app can use a suffix. So this implies that if required, one can expose each app with its own suffix or an inherited suffix. This might not be what the project wants, and I can appreciate that.
>
> I feel that in order to break that assumption safely old installations in the wild will need to be updated I feel. Otherwise a `pipx` upgrade can break existing environments. Again, a fair few assumptions here.

I think right now we don't want to support any suffixes besides the main package, (@cs01 can correct me).  Doing otherwise sounds too complicated and confusing.  Currently the only pipx command with a `--suffix` option is `install`, so only the main package can currently have a suffix, and we don't have to worry about injected packages or dependencies in the wild having a suffix.

Definitely you are correct, `upgrade`'s use of `expose_apps()` needs a `suffix` argument, but I would argue that we currently only suffix the apps from the main package (as is the current behavior.)

In general, so far in pipx, injected packages are definitely second-class citizens.  We are not implementing everything for them that we do for the main package, like suffixes on their apps.

Grafting the `--suffix` option onto pipx clearly had more side-effects than originally envisaged.  I'm just trying to figure out the least-bad way to leave the code after still fixing these bugs.  Viewing all the code paths at this point in time as possible may not make sense because we probably need some refactoring and rethinking to completely integrate suffixes.

Reply:
Did you actually see a failure with `upgrade` and `package = venv.pipx_metadata.main_package`, or are you basing this based just on the code?

Reply:
> That would be incorrect. What I was proposing was to obtain metadata on the main package was:

That was a typo on my side. My point still stands that this overrides a mandatory argument in the `upgrade` function. Even when a package is provided, we reset this to the `main_package`. If this is indeed the intended behaviour we should probably drop the `package` parameter and simply rely on the metadata of the `Venv`.

> I would argue that we currently only suffix the apps from the main package (as is the current behavior.)

I am happy to do that, but I just find that a step backwards, especially since there already is a TODO to support this.

> Viewing all the code paths at this point in time as possible may not make sense because we probably need some refactoring and rethinking to completely integrate suffixes.

I agree. However, this is not necessarily handling all code paths, but rather just the ones that seemed obvious due to the interface. And as I said, happy to reduce the scope of the PR.

Reply:
> package = venv.pipx_metadata.main_package
> package_metadata = venv.package_metadata[package]

Isin't `venv.pipx_metadata.main_package` a `PackageInfo` instance?

Reply:
> > package = venv.pipx_metadata.main_package
> > package_metadata = venv.package_metadata[package]
> 
> Isin't `venv.pipx_metadata.main_package` a `PackageInfo` instance?

Ah yes you are right.  Sorry I was just typing from my mind.

Reply:
As for `upgrade()`, right now the `venv_dir` and and `package` arguments I believe are more of an implementation detail that are supposed to track each other--`venv_dir` is just the full path and `package` is just the name of the package.

Reply:
Another possibility that could be considered is that 
`@property def package_metadata(self)` in class `Venv` returns a dict with package + package_suffix as the keys to the dict instead of just package.  Do you have any idea if that would help things?  Or would it mess up something else that depends on unsuffixed package names?

Reply:
I recall I had tried to use that as a fix initially; but there was something that felt harder than it should be and existing tests were failing. I can try that again see how far I can get with it. This simply avoids the loop in the cases where the main package is suffixed and is the one for which the metadata is being retrived. Although, overall I think it is probably not ideal as most lookups are package names, not name + suffix.

Additionally, while trying to implement your suggestions, I realiased that the `run_post_install_actions` will also have similar issues. SInce this is also called after inject with a non-main package. So it should handle both cases to support today's use cases. 

I could do something like this in places where a package is passed in. But then again this logic might be better placed in a method under `Venv`.

```py
    suffixed_package_name = f"{venv.pipx_metadata.main_package.package}{venv.pipx_metadata.main_package.suffix}"
    if package == suffixed_package_name:
        package = venv.pipx_metadata.main_package.package or package

    if package not in venv.package_metadata:
        raise PipxError(f"Package {red(bold(package))} not found in {venv.root}.")
```

Reply:
What if you modified `def package_metadata(self)` to only apply the suffix for the main_package?  The main package is already applied in a separate if-clause anyway.

Is there a way this breaks?

Reply:
I was playing around with the code a bit, and following seems to work at least with `pipx list`:
```
    @property
    def package_metadata(self) -> Dict[str, PackageInfo]:
        return_dict = self.pipx_metadata.injected_packages.copy()
        if self.pipx_metadata.main_package.package is not None:
            return_dict[
                self.pipx_metadata.main_package.package + self.pipx_metadata.main_package.suffix
            ] = self.pipx_metadata.main_package
        return return_dict
```
I used most of your `pipx_metadata_file.py` code.

I don't think we should try to apply the suffix to everything, just the main package and its apps.

Reply:
I need to double check how existing installs would be affected. Will do that when I get some time to look at this again. 

I suppose the policy question, even if not permanent, should be addressed before reducing the change scope.

Reply:
@itsayellow I have updated things a bit. The change defaults to venv's main package instead of venv root name. I have also removed the fallback loop to determine suffixed packaged.

Decided to retain the helper to abstract away some common functionality, and will also make maintenance easier in the future.

Let me know what you think about this version.